IO Read operation failed after deleting 2 segment store pods simultaneously

**Problem description**
I started long term IO operation with following configurations:

`IO workload (Total - 4 readers, 3 writers, 120 events/sec, ~ 100 KB/sec IO)`
While Read and Write operations were in-progress, deleted 2 Segment store pods out of 3. The pods restarted immediately but IO Read operation failed.

After Segment store  restart, following are the exceptions in IO client:

```
INFO  [2019-04-18 10:42:06,225] io.pravega.longevity.testmanager.manager.TestPerformanceCollector: Received Error From task taskID : io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.2.15 is not established.
java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.2.15 is not established.
        at java.util.concurrent.CompletableFuture.reportJoin(Unknown Source)
        at java.util.concurrent.CompletableFuture.join(Unknown Source)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.handleRequest(SegmentInputStreamImpl.java:137)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.read(SegmentInputStreamImpl.java:122)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.readEvent(EventSegmentReaderImpl.java:75)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.read(EventSegmentReaderImpl.java:62)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:115)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:88)
        at io.pravega.longevity.testworker.workers.readers.PravegaReader.start(PravegaReader.java:99)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.2.15 is not established.
        at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$null$6(Retry.java:245)
        at java.util.concurrent.CompletableFuture.uniExceptionally(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(Unknown Source)
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture.complete(Unknown Source)
        at io.pravega.common.concurrent.Futures.lambda$delayedFuture$19(Futures.java:516)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(Unknown Source)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
        ... 3 more
Caused by: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.2.15 is not established.
        at java.util.concurrent.CompletableFuture.encodeRelay(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniComposeStage(Unknown Source)
        at java.util.concurrent.CompletableFuture.thenCompose(Unknown Source)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.lambda$read$3(AsyncSegmentInputStreamImpl.java:188)
        at io.pravega.common.concurrent.Futures.lambda$delayedFuture$21(Futures.java:536)
        at java.util.concurrent.CompletableFuture.uniCompose(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
        ... 9 more
Caused by: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.2.15 is not established.
        at io.pravega.client.netty.impl.FlowHandler.getChannel(FlowHandler.java:121)
        at io.pravega.client.netty.impl.ClientConnectionImpl.sendAsync(ClientConnectionImpl.java:69)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.sendRequestOverConnection(AsyncSegmentInputStreamImpl.java:202)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.lambda$null$2(AsyncSegmentInputStreamImpl.java:188)
        ... 15 more

172.24.93.3 - - [18/Apr/2019:10:42:06 +0000] "POST /v1/longevity/error HTTP/1.1" 200 3791 "-" "Jersey/2.25.1 (HttpUrlConnection 1.8.0_131)" 5
INFO  [2019-04-18 10:42:06,835] io.pravega.longevity.testmanager.manager.TestPerformanceCollector: Received Error From task taskID : io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.2.15 is not established.
java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.2.15 is not established.
        at java.util.concurrent.CompletableFuture.reportJoin(Unknown Source)
        at java.util.concurrent.CompletableFuture.join(Unknown Source)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.handleRequest(SegmentInputStreamImpl.java:137)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.read(SegmentInputStreamImpl.java:122)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.readEvent(EventSegmentReaderImpl.java:75)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.read(EventSegmentReaderImpl.java:62)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:115)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:88)
        at io.pravega.longevity.testworker.workers.readers.PravegaReader.start(PravegaReader.java:99)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.2.15 is not established.
        at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$null$6(Retry.java:245)
        at java.util.concurrent.CompletableFuture.uniExceptionally(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(Unknown Source)
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture.complete(Unknown Source)
        at io.pravega.common.concurrent.Futures.lambda$delayedFuture$19(Futures.java:516)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(Unknown Source)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
        ... 3 more
Caused by: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.2.15 is not established.
        at java.util.concurrent.CompletableFuture.encodeRelay(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniComposeStage(Unknown Source)
        at java.util.concurrent.CompletableFuture.thenCompose(Unknown Source)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.lambda$read$3(AsyncSegmentInputStreamImpl.java:188)
        at io.pravega.common.concurrent.Futures.lambda$delayedFuture$21(Futures.java:536)
        at java.util.concurrent.CompletableFuture.uniCompose(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
        ... 9 more
Caused by: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.2.15 is not established.
        at io.pravega.client.netty.impl.FlowHandler.getChannel(FlowHandler.java:121)
        at io.pravega.client.netty.impl.ClientConnectionImpl.sendAsync(ClientConnectionImpl.java:69)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.sendRequestOverConnection(AsyncSegmentInputStreamImpl.java:202)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.lambda$null$2(AsyncSegmentInputStreamImpl.java:188)
        ... 15 more

172.24.93.3 - - [18/Apr/2019:10:42:06 +0000] "POST /v1/longevity/error HTTP/1.1" 200 3791 "-" "Jersey/2.25.1 (HttpUrlConnection 1.8.0_131)" 3
INFO  [2019-04-18 10:42:32,674] io.pravega.longevity.utils.PerformanceUtils: Got new data  Running for 1h 20m
INFO  [2019-04-18 10:42:32,674] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:203,417, events/sec:42, KB/sec:32.490234
INFO  [2019-04-18 10:42:32,674] io.pravega.longevity.utils.PerformanceUtils:  Transactions: started: 0, perSec:0.0, committed:0, failed:0, aborted:0, aborted but read: 0
INFO  [2019-04-18 10:42:32,674] io.pravega.longevity.utils.PerformanceUtils:  Blocked Waiting For ACK: times blocks: 0, average block time (ms): 0
INFO  [2019-04-18 10:42:32,674] io.pravega.longevity.utils.PerformanceUtils: Readers (2/4): events:405,210, events/sec:84, KB/sec:64.72168
```
IO Client exceptions:

```
2019-04-18 11:02:35,992 6010995 [epollEventLoopGroup-2-5] WARN  i.p.c.s.i.SegmentOutputStreamImpl - 7da43963-5855-4b18-a9ee-d69dc39cc490 Failed to connect: 
java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: io.netty.handler.codec.EncoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0
        at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniExceptionally(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(Unknown Source)
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source)
        at io.pravega.client.segment.impl.SegmentOutputStreamImpl$State.failConnection(SegmentOutputStreamImpl.java:202)
        at io.pravega.client.segment.impl.SegmentOutputStreamImpl$State.access$700(SegmentOutputStreamImpl.java:94)
        at io.pravega.client.segment.impl.SegmentOutputStreamImpl.failConnection(SegmentOutputStreamImpl.java:516)
        at io.pravega.client.segment.impl.SegmentOutputStreamImpl.access$500(SegmentOutputStreamImpl.java:68)
        at io.pravega.client.segment.impl.SegmentOutputStreamImpl$ResponseProcessor.lambda$appendSetup$1(SegmentOutputStreamImpl.java:366)
        at io.pravega.client.netty.impl.ClientConnectionImpl.lambda$sendAsync$1(ClientConnectionImpl.java:105)
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:511)
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:485)
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:424)
        at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:121)
        at io.netty.util.concurrent.PromiseCombiner.tryPromise(PromiseCombiner.java:126)
        at io.netty.util.concurrent.PromiseCombiner.finish(PromiseCombiner.java:121)
        at io.pravega.client.netty.impl.ClientConnectionImpl.sendAsync(ClientConnectionImpl.java:107)
        at io.pravega.client.segment.impl.SegmentOutputStreamImpl$ResponseProcessor.appendSetup(SegmentOutputStreamImpl.java:362)
        at io.pravega.shared.protocol.netty.WireCommands$AppendSetup.process(WireCommands.java:666)
        at io.pravega.shared.protocol.netty.ReplyProcessor.process(ReplyProcessor.java:20)
        at io.pravega.client.netty.impl.FlowHandler.lambda$channelRead$2(FlowHandler.java:195)
        at java.util.Optional.ifPresent(Unknown Source)
        at io.pravega.client.netty.impl.FlowHandler.channelRead(FlowHandler.java:193)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
        at io.pravega.shared.protocol.netty.ExceptionLoggingHandler.channelRead(ExceptionLoggingHandler.java:37)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:421)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:321)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.shared.protocol.netty.ConnectionFailedException: io.netty.handler.codec.EncoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0
        ... 42 common frames omitted
Caused by: io.netty.handler.codec.EncoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0
        at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:125)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:816)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:723)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:704)
        at io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1056)
        at io.netty.channel.AbstractChannel.write(AbstractChannel.java:290)
        at io.pravega.client.netty.impl.ClientConnectionImpl.sendAsync(ClientConnectionImpl.java:99)
        ... 34 common frames omitted
Caused by: io.netty.util.IllegalReferenceCountException: refCnt: 0
        at io.netty.buffer.AbstractByteBuf.ensureAccessible(AbstractByteBuf.java:1446)
        at io.netty.buffer.AbstractByteBuf.slice(AbstractByteBuf.java:1216)
        at io.netty.buffer.AbstractByteBuf.slice(AbstractByteBuf.java:1206)
        at io.pravega.shared.protocol.netty.CommandEncoder.encode(CommandEncoder.java:108)
        at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:107)
        ... 42 common frames omitted
2019-04-18 11:02:35,992 6010995 [epollEventLoopGroup-2-5] DEBUG io.pravega.common.util.Retry - Retrying command. Retry #186, timestamp=2019-04-18T11:02:35.992Z
```
Environment details: PKS / K8 with medium cluster:

```
3 master nodes @ large.cpu (4 CPU, 4 GB Ram, 16 GB Disk)
5 worker nodes @ xlarge.cpu(8 cpu, 8 GB Ram, 32 GB Disk)
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
Deployment Details:
```
Pravega-Operator image: pravega/pravega-operator:0.3.2
Pravega build used : 0.5.0-2222.8ba5431
Zookeeper-operator image: pravega/zookeeper-operator:0.2.1
```

Following are the Cluster details:
```
root@osboxes:/home/osboxes# kubectl get po
NAME                                                 READY     STATUS    RESTARTS   AGE
pravega-bookie-0                                     1/1       Running   0          2d
pravega-bookie-1                                     1/1       Running   0          4d
pravega-bookie-2                                     1/1       Running   0          4d
pravega-bookie-3                                     1/1       Running   0          4d
pravega-operator-779879b48-6j28t                     1/1       Running   0          4d
pravega-pravega-controller-84948d8895-k6mrm          1/1       Running   1          2d
pravega-pravega-controller-84948d8895-krxpq          1/1       Running   2          4d
pravega-pravega-segmentstore-0                       1/1       Running   0          2d
pravega-pravega-segmentstore-1                       1/1       Running   0          2d
pravega-pravega-segmentstore-2                       1/1       Running   0          2d
pravega-zk-0                                         1/1       Running   0          2d
pravega-zk-1                                         0/1       Running   0          4d
pravega-zk-2                                         0/1       Running   0          4d
pravega-zk-3                                         0/1       Running   0          4d
shaka-zulu-nfs-client-provisioner-59d7f8f84c-942d2   1/1       Running   0          4d
zookeeper-operator-685bfcbbc5-wzchs                  1/1       Running   0          4d

```
Detailed logs from Pravega all pods are available in `CompelteLogs.zip`
Attaching complete 
[CompleteLogs.zip](https://github.com/pravega/pravega/files/3102392/CompleteLogs.zip)
Logs for all the components

@Richa-Srivastava @sumit-bm thanks for sharing the logs. 
My current analysis is as follows.

1) Error from test
```
INFO  [2019-04-23 12:04:25,526] io.pravega.longevity.testmanager.manager.TestPerformanceCollector: Received Error From task taskID : io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.75.11 is not established.
java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.75.11 is not established.
        at java.util.concurrent.CompletableFuture.reportJoin(Unknown Source)
        at java.util.concurrent.CompletableFuture.join(Unknown Source)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.handleRequest(SegmentInputStreamImpl.java:137)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.read(SegmentInputStreamImpl.java:122)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.readEvent(EventSegmentReaderImpl.java:75)
        at io.pravega.client.segment.impl.EventSegmentReaderImpl.read(EventSegmentReaderImpl.java:62)
        at io.pravega.client.segment.impl.EventSegmentReader.read(EventSegmentReader.java:53)
        at io.pravega.client.state.impl.RevisionedStreamClientImpl$StreamIterator.next(RevisionedStreamClientImpl.java:150)
        at io.pravega.client.state.impl.RevisionedStreamClientImpl$StreamIterator.next(RevisionedStreamClientImpl.java:125)
        at io.pravega.client.state.impl.StateSynchronizerImpl.fetchUpdates(StateSynchronizerImpl.java:80)
        at io.pravega.client.stream.impl.ReaderGroupStateManager.fetchUpdatesIfNeeded(ReaderGroupStateManager.java:284)
        at io.pravega.client.stream.impl.ReaderGroupStateManager.getCheckpoint(ReaderGroupStateManager.java:377)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.updateGroupStateIfNeeded(EventStreamReaderImpl.java:171)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:103)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:88)
        at io.pravega.longevity.testworker.workers.readers.PravegaReader.start(PravegaReader.java:99)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.75.11 is not established.
        at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$null$6(Retry.java:245)
        at java.util.concurrent.CompletableFuture.uniExceptionally(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(Unknown Source)
        at java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
        at java.util.concurrent.CompletableFuture.complete(Unknown Source)
        at io.pravega.common.concurrent.Futures.lambda$delayedFuture$19(Futures.java:516)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(Unknown Source)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
        ... 3 more
Caused by: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.75.11 is not established.
        at java.util.concurrent.CompletableFuture.encodeRelay(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniComposeStage(Unknown Source)
        at java.util.concurrent.CompletableFuture.thenCompose(Unknown Source)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.lambda$read$3(AsyncSegmentInputStreamImpl.java:188)
        at io.pravega.common.concurrent.Futures.lambda$delayedFuture$21(Futures.java:536)
        at java.util.concurrent.CompletableFuture.uniCompose(Unknown Source)
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source)
        ... 9 more
Caused by: io.pravega.shared.protocol.netty.ConnectionFailedException: Connection to 172.24.75.11 is not established.
        at io.pravega.client.netty.impl.FlowHandler.getChannel(FlowHandler.java:121)
        at io.pravega.client.netty.impl.ClientConnectionImpl.sendAsync(ClientConnectionImpl.java:69)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.sendRequestOverConnection(AsyncSegmentInputStreamImpl.java:202)
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.lambda$null$2(AsyncSegmentInputStreamImpl.java:188)
        ... 15 more
```

2) From the logs it can been seen that `172.24.75.11` also corresponds to segmentStore-2 (ref : describe_pod_pravega-pravega-segmentstore-2_output) 
Since this instance of segment store starts @ ``12:04:40,585`` I am assuming that this IP is getting reused.
Also, `pravega-segmentstore-1` starts @ `12:04:12,619 ` I am assuming that segment stores (1 and 2) were  stopped at a time stamp before this.
3) The stack trace of the issue seems similar to the issue fixed by PR: #3639.

--- a/client/src/main/java/io/pravega/client/ClientConfig.java
+++ b/client/src/main/java/io/pravega/client/ClientConfig.java
@@ -59,12 +59,12 @@ public class ClientConfig implements Serializable {
     /**
      * If the flag {@link #isEnableTls()}  is set, this flag decides whether to enable host name validation or not.
      */
-    private boolean validateHostName;
+    private final boolean validateHostName;
 
     /**
      * Maximum number of connections per Segment store.
      */
-    private int maxConnectionsPerSegmentStore;
+    private final int maxConnectionsPerSegmentStore;
 
     public boolean isEnableTls() {
         String scheme = this.controllerURI.getScheme();

--- a/client/src/main/java/io/pravega/client/netty/impl/Connection.java
+++ b/client/src/main/java/io/pravega/client/netty/impl/Connection.java
@@ -9,9 +9,7 @@
  */
 package io.pravega.client.netty.impl;
 
-import io.pravega.common.concurrent.Futures;
 import io.pravega.shared.protocol.netty.PravegaNodeUri;
-import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import lombok.Data;
 
@@ -22,16 +20,18 @@ import lombok.Data;
 @Data
 public class Connection {
     private final PravegaNodeUri uri;
-    private final CompletableFuture<FlowHandler> flowHandler;
+    private final FlowHandler flowHandler;
     
     /**
-     * Returns the number of open flows on this connection, if the connection has been established.
+     * A future that completes when the connection is first established.
      */
-    public Optional<Integer> getFlowCount() {
-        if (!Futures.isSuccessful(flowHandler)) {
-            return Optional.empty();
-        }
-        return Optional.of(flowHandler.join().getOpenFlowCount());
+    private final CompletableFuture<Void> connected;
+    
+    /**
+     * Returns the number of open flows on this connection. 
+     */
+    public int getFlowCount() {
+        return flowHandler.getOpenFlowCount();
     }
 }
 

--- a/client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java
+++ b/client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java
@@ -72,8 +72,8 @@ public class ConnectionPoolImpl implements ConnectionPool {
     private static final Comparator<Connection> COMPARATOR = new Comparator<Connection>() {
         @Override
         public int compare(Connection c1, Connection c2) {
-            int v1 = c1.getFlowCount().orElse(Integer.MAX_VALUE);
-            int v2 = c2.getFlowCount().orElse(Integer.MAX_VALUE);
+            int v1 = Futures.isSuccessful(c1.getConnected()) ? c1.getFlowCount() : Integer.MAX_VALUE;
+            int v2 = Futures.isSuccessful(c2.getConnected()) ? c2.getFlowCount() : Integer.MAX_VALUE;
             return Integer.compare(v1, v2);
         }
     }; 
@@ -105,8 +105,7 @@ public class ConnectionPoolImpl implements ConnectionPool {
         // remove connections for which the underlying network connection is disconnected.
         List<Connection> prunedConnectionList = connectionList.stream().filter(connection -> {
             // Filter out Connection objects which have been completed exceptionally or have been disconnected.
-            CompletableFuture<FlowHandler> r = connection.getFlowHandler();
-            return !r.isDone() || (Futures.isSuccessful(r) && r.join().isConnectionEstablished());
+            return !connection.getConnected().isDone() || (Futures.isSuccessful(connection.getConnected()) && connection.getFlowHandler().isConnectionEstablished());
         }).collect(Collectors.toList());
         log.debug("List of connections to {} that can be used: {}", location, prunedConnectionList);
 
@@ -115,19 +114,20 @@ public class ConnectionPoolImpl implements ConnectionPool {
 
         final Connection connection;
         if (suggestedConnection.isPresent() && (prunedConnectionList.size() >= clientConfig.getMaxConnectionsPerSegmentStore() || isUnused(suggestedConnection.get()))) {
-            log.debug("Reusing connection: {}", suggestedConnection.get());
-            Connection oldConnection = suggestedConnection.get();
-            prunedConnectionList.remove(oldConnection);
-            connection = new Connection(oldConnection.getUri(), oldConnection.getFlowHandler());
+            log.info("Reusing connection: {}", suggestedConnection.get());
+            connection = suggestedConnection.get();
         } else {
             // create a new connection.
-            log.debug("Creating a new connection to {}", location);
-            CompletableFuture<FlowHandler> flowHandlerFuture = establishConnection(location);
-            connection = new Connection(location, flowHandlerFuture);
+            log.info("Creating a new connection to {}", location);
+            final AppendBatchSizeTracker batchSizeTracker = new AppendBatchSizeTrackerImpl();
+            final FlowHandler handler = new FlowHandler(location.getEndpoint(), batchSizeTracker);
+            CompletableFuture<Void> establishedFuture = establishConnection(location, handler);
+            connection = new Connection(location, handler, establishedFuture);
+            prunedConnectionList.add(connection);
         }
-        prunedConnectionList.add(connection);
+        ClientConnection result = connection.getFlowHandler().createFlow(flow, rp);
         connectionMap.put(location, prunedConnectionList);
-        return connection.getFlowHandler().thenApply(flowHandler -> flowHandler.createFlow(flow, rp));
+        return connection.getConnected().thenApply(v -> result);
     }
 
     @Override
@@ -137,25 +137,29 @@ public class ConnectionPoolImpl implements ConnectionPool {
         Exceptions.checkNotClosed(closed.get(), this);
 
         // create a new connection.
-        CompletableFuture<FlowHandler> flowHandlerFuture = establishConnection(location);
-        Connection connection = new Connection(location, flowHandlerFuture);
-        return connection.getFlowHandler().thenApply(flowHandler -> flowHandler.createConnectionWithFlowDisabled(rp));
+        final AppendBatchSizeTracker batchSizeTracker = new AppendBatchSizeTrackerImpl();
+        final FlowHandler handler = new FlowHandler(location.getEndpoint(), batchSizeTracker);
+        CompletableFuture<Void> connectedFuture = establishConnection(location, handler);
+        Connection connection = new Connection(location, handler, connectedFuture);
+        ClientConnection result = connection.getFlowHandler().createConnectionWithFlowDisabled(rp);
+        return connectedFuture.thenApply(v -> result);
     }
 
     private boolean isUnused(Connection connection) {
-        return connection.getFlowCount().isPresent() && connection.getFlowCount().get() == 0;
+        return Futures.isSuccessful(connection.getConnected()) && connection.getFlowCount() == 0;
     }
 
     /**
      * Used only for testing.
      */
     @VisibleForTesting
+    @Synchronized
     public void pruneUnusedConnections() {
         for (List<Connection> connections : connectionMap.values()) {
             for (Iterator<Connection> iterator = connections.iterator(); iterator.hasNext();) {
                 Connection connection = iterator.next();
                 if (isUnused(connection)) {
-                    connection.getFlowHandler().join().close();
+                    connection.getFlowHandler().close();
                     iterator.remove();
                 }
             }
@@ -178,16 +182,14 @@ public class ConnectionPoolImpl implements ConnectionPool {
     /**
      * Establish a new connection to the Pravega Node.
      * @param location The Pravega Node Uri
+     * @param handler The flow handler for the connection
      * @return A future, which completes once the connection has been established, returning a FlowHandler that can be used to create
      * flows on the connection.
      */
-    private CompletableFuture<FlowHandler> establishConnection(PravegaNodeUri location) {
-        final AppendBatchSizeTracker batchSizeTracker = new AppendBatchSizeTrackerImpl();
-        final FlowHandler handler = new FlowHandler(location.getEndpoint(), batchSizeTracker);
-        final Bootstrap b = getNettyBootstrap().handler(getChannelInitializer(location, batchSizeTracker, handler));
-
+    private CompletableFuture<Void> establishConnection(PravegaNodeUri location, FlowHandler handler) {  
+        final Bootstrap b = getNettyBootstrap().handler(getChannelInitializer(location, handler));
         // Initiate Connection.
-        final CompletableFuture<FlowHandler> connectionComplete = new CompletableFuture<>();
+        final CompletableFuture<Void> connectionComplete = new CompletableFuture<>();
         try {
             b.connect(location.getEndpoint(), location.getPort()).addListener((ChannelFutureListener) future -> {
                 if (future.isSuccess()) {
@@ -196,7 +198,7 @@ public class ConnectionPoolImpl implements ConnectionPool {
                     log.debug("Connect operation completed for channel:{}, local address:{}, remote address:{}",
                               ch.id(), ch.localAddress(), ch.remoteAddress());
                     channelGroup.add(ch); // Once a channel is closed the channel group implementation removes it.
-                    connectionComplete.complete(handler);
+                    connectionComplete.complete(null);
                 } else {
                     connectionComplete.completeExceptionally(new ConnectionFailedException(future.cause()));
                 }
@@ -208,7 +210,7 @@ public class ConnectionPoolImpl implements ConnectionPool {
         final CompletableFuture<Void> channelRegisteredFuture = new CompletableFuture<>(); //to track channel registration.
         handler.completeWhenRegistered(channelRegisteredFuture);
 
-        return connectionComplete.thenCombine(channelRegisteredFuture, (flowHandler, v) -> flowHandler);
+        return CompletableFuture.allOf(connectionComplete, channelRegisteredFuture);
     }
 
     /**
@@ -226,7 +228,6 @@ public class ConnectionPoolImpl implements ConnectionPool {
      * Create a Channel Initializer which is to to setup {@link ChannelPipeline}.
      */
     private ChannelInitializer<SocketChannel> getChannelInitializer(final PravegaNodeUri location,
-                                                                    final AppendBatchSizeTracker batchSizeTracker,
                                                                     final FlowHandler handler) {
         final SslContext sslCtx = getSslContext();
 
@@ -247,7 +248,7 @@ public class ConnectionPoolImpl implements ConnectionPool {
                 }
                 p.addLast(
                         new ExceptionLoggingHandler(location.getEndpoint()),
-                        new CommandEncoder(batchSizeTracker),
+                        new CommandEncoder(handler.getBatchSizeTracker()),
                         new LengthFieldBasedFrameDecoder(WireCommands.MAX_WIRECOMMAND_SIZE, 4, 4),
                         new CommandDecoder(),
                         handler);

--- a/client/src/main/java/io/pravega/client/netty/impl/FlowHandler.java
+++ b/client/src/main/java/io/pravega/client/netty/impl/FlowHandler.java
@@ -42,6 +42,7 @@ public class FlowHandler extends ChannelInboundHandlerAdapter implements AutoClo
     private final AtomicReference<ScheduledFuture<?>> keepAliveFuture = new AtomicReference<>();
     private final AtomicBoolean recentMessage = new AtomicBoolean(false);
     private final AtomicBoolean closed = new AtomicBoolean(false);
+    @Getter
     private final AppendBatchSizeTracker batchSizeTracker;
     private final ReusableFutureLatch<Void> registeredFutureLatch = new ReusableFutureLatch<>();
     @VisibleForTesting

--- a/client/src/test/java/io/pravega/client/netty/impl/ConnectionPoolingTest.java
+++ b/client/src/test/java/io/pravega/client/netty/impl/ConnectionPoolingTest.java
@@ -35,6 +35,7 @@ import io.pravega.shared.protocol.netty.ConnectionFailedException;
 import io.pravega.shared.protocol.netty.FailingReplyProcessor;
 import io.pravega.shared.protocol.netty.PravegaNodeUri;
 import io.pravega.shared.protocol.netty.WireCommands;
+import io.pravega.test.common.AssertExtensions;
 import io.pravega.test.common.TestUtils;
 import java.io.File;
 import java.net.URI;
@@ -157,13 +158,16 @@ public class ConnectionPoolingTest {
     }
 
     @Test
-    public void testConnectionPooling() throws ConnectionFailedException, InterruptedException {
+    public void testConnectionPooling() throws Exception {
+        ClientConfig clientConfig = ClientConfig.builder()
+                .controllerURI(URI.create((this.ssl ? "tls://" : "tcp://")
+                                          + "localhost"))
+                .trustStore("../config/cert.pem")
+                .maxConnectionsPerSegmentStore(1)
+                .build();
+        ConnectionPoolImpl connectionPool = new ConnectionPoolImpl(clientConfig);
         @Cleanup
-        ConnectionFactoryImpl factory = new ConnectionFactoryImpl(ClientConfig.builder()
-                                                                              .controllerURI(URI.create((this.ssl ? "tls://" : "tcp://") + "localhost"))
-                                                                              .trustStore("../config/cert.pem")
-                                                                              .maxConnectionsPerSegmentStore(1)
-                                                                              .build());
+        ConnectionFactoryImpl factory = new ConnectionFactoryImpl(clientConfig, connectionPool, 1);
 
         ArrayBlockingQueue<WireCommands.SegmentRead> msgRead = new ArrayBlockingQueue<>(10);
         FailingReplyProcessor rp = new FailingReplyProcessor() {
@@ -228,17 +232,24 @@ public class ConnectionPoolingTest {
         // close connection1
         connection1.close();
         assertThrows(ConnectionFailedException.class, () -> connection1.send(readRequestGenerator.apply(flow2.asLong())));
+        AssertExtensions.assertEventuallyEquals(0, () -> {
+            connectionPool.pruneUnusedConnections();
+            return factory.getActiveChannelCount();
+        }, 10000);
 
     }
 
     @Test
-    public void testConcurrentRequests() throws ConnectionFailedException, InterruptedException {
+    public void testConcurrentRequests() throws Exception {
+        ClientConfig clientConfig = ClientConfig.builder()
+                                                .controllerURI(URI.create((this.ssl ? "tls://" : "tcp://")
+                                                        + "localhost"))
+                                                .trustStore("../config/cert.pem")
+                                                .maxConnectionsPerSegmentStore(1)
+                                                .build();
+        ConnectionPoolImpl connectionPool = new ConnectionPoolImpl(clientConfig);
         @Cleanup
-        ConnectionFactoryImpl factory = new ConnectionFactoryImpl(ClientConfig.builder()
-                                                                              .controllerURI(URI.create((this.ssl ? "tls://" : "tcp://") + "localhost"))
-                                                                              .trustStore("../config/cert.pem")
-                                                                              .maxConnectionsPerSegmentStore(1)
-                                                                              .build());
+        ConnectionFactoryImpl factory = new ConnectionFactoryImpl(clientConfig, connectionPool, 1);
 
         ArrayBlockingQueue<WireCommands.SegmentRead> msgRead = new ArrayBlockingQueue<>(10);
         FailingReplyProcessor rp = new FailingReplyProcessor() {
@@ -264,12 +275,10 @@ public class ConnectionPoolingTest {
         };
 
         Flow flow1 = new Flow(1, 0);
-        @Cleanup
         ClientConnection connection1 = factory.establishConnection(flow1, new PravegaNodeUri("localhost", port), rp).join();
 
         // create a second connection, since the max number of connections is 1 this should reuse the same connection.
         Flow flow2 = new Flow(2, 0);
-        @Cleanup
         ClientConnection connection2 = factory.establishConnection(flow2, new PravegaNodeUri("localhost", port), rp).join();
 
         connection1.send(readRequestGenerator.apply(flow1.asLong()));
@@ -280,6 +289,13 @@ public class ConnectionPoolingTest {
         msgs.add(msgRead.take());
         assertTrue(msgs.contains(readResponseGenerator.apply(flow1.asLong())));
         assertTrue(msgs.contains(readResponseGenerator.apply(flow1.asLong())));
+        assertEquals(1, factory.getActiveChannelCount());
+        connection1.close();
+        connection2.close();
+        AssertExtensions.assertEventuallyEquals(0, () -> {
+            connectionPool.pruneUnusedConnections();
+            return factory.getActiveChannelCount();
+        }, 10000);
     }
 
 }

--- a/test/integration/src/test/java/io/pravega/test/integration/endtoendtest/EndToEndChannelLeakTest.java
+++ b/test/integration/src/test/java/io/pravega/test/integration/endtoendtest/EndToEndChannelLeakTest.java
@@ -40,7 +40,6 @@ import io.pravega.test.integration.demo.ControllerWrapper;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 import lombok.Cleanup;
 import lombok.extern.slf4j.Slf4j;
@@ -49,10 +48,10 @@ import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 
+import static io.pravega.test.common.AssertExtensions.assertEventuallyEquals;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
-import static io.pravega.test.common.AssertExtensions.assertEventuallyEquals;
 
 @Slf4j
 public class EndToEndChannelLeakTest {
@@ -75,7 +74,7 @@ public class EndToEndChannelLeakTest {
 
     @Before
     public void setUp() throws Exception {
-        executor = Executors.newSingleThreadScheduledExecutor();
+        executor = new InlineExecutor();
         zkTestServer = new TestingServerStarter().start();
 
         serviceBuilder = ServiceBuilder.newInMemoryBuilder(ServiceBuilderConfig.getDefaultConfig());
@@ -103,7 +102,7 @@ public class EndToEndChannelLeakTest {
     @Test(timeout = 30000)
     public void testDetectChannelLeakSegmentSealedPooled() throws Exception {
         StreamConfiguration config = StreamConfiguration.builder()
-                                                        .scalingPolicy(ScalingPolicy.byEventRate(10, 2, 1))
+                                                        .scalingPolicy(ScalingPolicy.fixed(1))
                                                         .build();
         Controller controller = controllerWrapper.getController();
         controllerWrapper.getControllerService().createScope(SCOPE).get();
@@ -158,15 +157,15 @@ public class EndToEndChannelLeakTest {
         event = reader1.readNextEvent(10000);
         assertNotNull(event.getEvent());
 
-        assertEventuallyEquals(5, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(5, connectionPool);
 
         event = reader1.readNextEvent(10000);
         assertNotNull(event.getEvent());
-        assertEventuallyEquals(5, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(5, connectionPool);
 
         event = reader1.readNextEvent(10000);
         assertNotNull(event.getEvent());
-        assertEventuallyEquals(5, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(5, connectionPool);
     }
 
     @Test(timeout = 30000)
@@ -180,12 +179,13 @@ public class EndToEndChannelLeakTest {
         Controller controller = controllerWrapper.getController();
         controllerWrapper.getControllerService().createScope(SCOPE).get();
         controller.createStream(SCOPE, STREAM_NAME, config).get();
+        ConnectionPoolImpl connectionPool = new ConnectionPoolImpl(clientConfig);
         @Cleanup
-        ConnectionFactoryImpl connectionFactory = new ConnectionFactoryImpl(clientConfig);
+        ConnectionFactoryImpl connectionFactory = new ConnectionFactoryImpl(clientConfig, connectionPool, executor);
         @Cleanup
         ClientFactoryImpl clientFactory = new ClientFactoryImpl(SCOPE, controller, connectionFactory);
         int expectedChannelCount = 0; // open socket count.
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(expectedChannelCount, connectionPool);
         
         //Create a writer and write an event.
         @Cleanup
@@ -194,12 +194,12 @@ public class EndToEndChannelLeakTest {
         writer.writeEvent("0", "zero").get();
 
         expectedChannelCount += 1; // connection to segment 0.
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(expectedChannelCount, connectionPool);
         
         @Cleanup
         ReaderGroupManager groupManager = new ReaderGroupManagerImpl(SCOPE, controller, clientFactory,
                 connectionFactory);
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT); // no changes expected.
+        assertChannelCount(expectedChannelCount, connectionPool); // no changes expected.
       
         groupManager.createReaderGroup(READER_GROUP, ReaderGroupConfig.builder().disableAutomaticCheckpoints()
                 .stream(Stream.of(SCOPE, STREAM_NAME)).build());
@@ -215,7 +215,7 @@ public class EndToEndChannelLeakTest {
         assertNotNull(event);
         assertEquals("zero", event.getEvent());
         //Connection to segment 0 does not cause an increase in number of open connections since we have reached the maxConnection count.
-        assertEventuallyEquals(5, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(5, connectionPool);
 
         // scale
         Stream stream = new StreamImpl(SCOPE, STREAM_NAME);
@@ -226,7 +226,7 @@ public class EndToEndChannelLeakTest {
         Boolean result = controller.scaleStream(stream, Collections.singletonList(0L), map, executor).getFuture().get();
         assertTrue(result);
         //No changes to the channel count.
-        assertEventuallyEquals(5, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(5, connectionPool);
 
         //Write more events.
         writer.writeEvent("1", "one").get();
@@ -238,7 +238,7 @@ public class EndToEndChannelLeakTest {
 
         //2 new flows  are opened.(+3 connections to the segments 1,2,3 after scale by the writer,
         // -1 flow to segment 0 which is sealed.)
-        assertEventuallyEquals(5, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(5, connectionPool);
 
         //Add a new reader
         @Cleanup
@@ -251,14 +251,14 @@ public class EndToEndChannelLeakTest {
 
         //+1 flow (-1 since segment 0 of stream is sealed + 2 connections to two segments of stream (there are
         // 2 readers and 3 segments and the reader1 will be assigned 2 segments))
-        assertEventuallyEquals(5, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(5, connectionPool);
 
         event = reader2.readNextEvent(10000);
         assertNotNull(event.getEvent());
 
         //+1 flow (a new flow to the remaining stream segment)
         expectedChannelCount += 1;
-        assertEventuallyEquals(5, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(5, connectionPool);
     }
     
     @Test(timeout = 30000)
@@ -296,7 +296,8 @@ public class EndToEndChannelLeakTest {
                 ReaderConfig.builder().build());
         //Write an event.
         writer.writeEvent("0", "zero").get();
-
+        
+        connectionPool.pruneUnusedConnections();
         int channelCount = connectionFactory.getActiveChannelCount(); //store the open channel count before reading.
 
         //Read an event.
@@ -304,7 +305,7 @@ public class EndToEndChannelLeakTest {
         assertEquals("zero", event.getEvent());
 
         channelCount += 1;
-        assertEventuallyEquals(channelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(channelCount, connectionPool);
         
         // scale
         Stream stream = new StreamImpl(SCOPE, SCOPE);
@@ -321,35 +322,31 @@ public class EndToEndChannelLeakTest {
         writer.writeEvent("1", "three").get();
         
         channelCount += 2; //from the new segments and close of old segment
-        connectionPool.pruneUnusedConnections();
-        assertEventuallyEquals(channelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(channelCount, connectionPool);
 
         event = reader1.readNextEvent(10000);
         assertNotNull(event.getEvent());
         //Number of sockets will increase by 3 for the new segments and decrease by 1 from the old segment 
         channelCount += 2;
-        connectionPool.pruneUnusedConnections();
-        assertEventuallyEquals(channelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(channelCount, connectionPool);
 
         event = reader1.readNextEvent(10000);
         assertNotNull(event.getEvent());
         //no changes to socket count.
-        assertEventuallyEquals(channelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(channelCount, connectionPool);
 
         event = reader1.readNextEvent(10000);
         assertNotNull(event.getEvent());
         //no changes to socket count.
-        assertEventuallyEquals(channelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(channelCount, connectionPool);
         
         reader1.close();
         //3 from segments 4 from group state.
         channelCount -= 7;
-        connectionPool.pruneUnusedConnections();
-        assertEventuallyEquals(channelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(channelCount, connectionPool);
         groupManager.close();
         writer.close();
-        connectionPool.pruneUnusedConnections();
-        assertEventuallyEquals(0, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(0, connectionPool);
 
     }
 
@@ -370,7 +367,7 @@ public class EndToEndChannelLeakTest {
         @Cleanup
         ClientFactoryImpl clientFactory = new ClientFactoryImpl(SCOPE, controller, connectionFactory);
         int expectedChannelCount = 0; // open socket count.
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(expectedChannelCount, connectionPool);
         
         //Create a writer and write an event.
         @Cleanup
@@ -379,12 +376,12 @@ public class EndToEndChannelLeakTest {
         writer.writeEvent("0", "zero").get();
 
         expectedChannelCount += 1; // connection to segment 0.
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(expectedChannelCount, connectionPool);
         
         @Cleanup
         ReaderGroupManager groupManager = new ReaderGroupManagerImpl(SCOPE, controller, clientFactory,
                 connectionFactory);
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT); // no changes expected.
+        assertChannelCount(expectedChannelCount, connectionPool); // no changes expected.
       
         groupManager.createReaderGroup(READER_GROUP, ReaderGroupConfig.builder().disableAutomaticCheckpoints()
                 .stream(Stream.of(SCOPE, STREAM_NAME)).build());
@@ -401,7 +398,7 @@ public class EndToEndChannelLeakTest {
         expectedChannelCount += 1;
         assertNotNull(event);
         assertEquals("zero", event.getEvent());
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(expectedChannelCount, connectionPool);
 
         // scale
         Stream stream = new StreamImpl(SCOPE, STREAM_NAME);
@@ -412,7 +409,7 @@ public class EndToEndChannelLeakTest {
         Boolean result = controller.scaleStream(stream, Collections.singletonList(0L), map, executor).getFuture().get();
         assertTrue(result);
         //No changes to the channel count.
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(expectedChannelCount, connectionPool);
 
         //Write more events.
         writer.writeEvent("1", "one").get();
@@ -425,8 +422,7 @@ public class EndToEndChannelLeakTest {
         //2 new flows  are opened.(+3 connections to the segments 1,2,3 after scale by the writer,
         // -1 flow to segment 0 which is sealed.)
         expectedChannelCount += 2;
-        connectionPool.pruneUnusedConnections();
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(expectedChannelCount, connectionPool);
 
         //Add a new reader
         @Cleanup
@@ -434,21 +430,28 @@ public class EndToEndChannelLeakTest {
                 ReaderConfig.builder().build());
         //Creating a reader spawns a revisioned stream client which opens 4 flows ( read, write, metadataClient and conditionalUpdates).
         expectedChannelCount += 4;
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(expectedChannelCount, connectionPool);
         event = reader1.readNextEvent(10000);
         assertNotNull(event.getEvent());
 
         //+1 flow (-1 since segment 0 of stream is sealed + 2 connections to two segments of stream (there are
         // 2 readers and 3 segments and the reader1 will be assigned 2 segments))
         expectedChannelCount += 1;
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(expectedChannelCount, connectionPool);
 
         event = reader2.readNextEvent(10000);
         assertNotNull(event.getEvent());
 
         //+1 flow (a new flow to the remaining stream segment)
         expectedChannelCount += 1;
-        assertEventuallyEquals(expectedChannelCount, () -> connectionFactory.getActiveChannelCount(), ASSERT_TIMEOUT);
+        assertChannelCount(expectedChannelCount, connectionPool);
+    }
+    
+    private void assertChannelCount(int expectedChannelCount, ConnectionPoolImpl connectionPool) throws Exception {
+        assertEventuallyEquals(expectedChannelCount, () -> {
+            connectionPool.pruneUnusedConnections();            
+            return connectionPool.getActiveChannelCount();
+        }, ASSERT_TIMEOUT);
     }
     
 }

